### PR TITLE
ROX-32259: Use real package data in fake workloads (shared pkg)

### PR DESCRIPTION
**Pull request by vikin91**
## Description

Replaces: #18171

This PR improves fake VM workload generation for load testing by:

1. **Using real RHEL 9 package data** - Replaced synthetic package generation with 509 real packages from a RHEL 9 system, including correct repository IDs (`rhel-9-for-x86_64-baseos-rpms`, `rhel-9-for-x86_64-appstream-rpms`) and proper CPEs.

2. **Flexible load generation** - Support for requesting more packages/repos than available in the fixture extracted from real RHEL 9 system:
   - `numPackages > 509`: duplicates packages with unique IDs
   - `numRepos > 2`: creates synthetic repos for load testing

3. **The load generation code is in shared pkg** - so that it can be used in multiple places without duplication.

**Before**: Fake index reports used generated package names like `pkg-template-4-608` with placeholder CPEs, causing scanner not scanning the package and returning immediately empty result to Central.

**After**: Reports use real RHEL 9 package names like `glibc`, `kernel`, `systemd` with valid repos and CPEs, so that scanner-v4-matcher is doing the real matching work and takes part in processing of the generated workload.

## User-facing documentation

- [x] [CHANGELOG.md](https://github.com/stackrox/stackrox/blob/master/CHANGELOG.md) is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

## Testing and quality

- [x] the change is production ready: the change is [GA](https://github.com/stackrox/stackrox/blob/master/PR_GA.md), or otherwise the functionality is gated by a [feature flag](https://github.com/stackrox/stackrox/blob/master/pkg/features/README.md)
- [ ] CI results are [inspected](https://docs.google.com/document/d/1d5ga073jkv4CO1kAJqp8MPGpC6E1bwyrCGZ7S5wKg3w/edit?tab=t.0#heading=h.w4ercgtcg0xp)

### Automated testing

- [x] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [x] modified existing tests

### How I validated my change

1. Ran unit tests locally:
   ```
   go test ./sensor/kubernetes/fake/... -v
   ```
   All tests pass including new tests for `validateVMWorkload`, `normalizeRPMVersion`, `selectPackageIndices`, and `buildRepositories`.

2. CI validation.

3. Manual validation was performed during development by running fake sensor with VM workloads and verifying that scanner is scanning the packages and returns vulnerabilities (where applicable).

Screenshot 2025-12-10 at 13 29 05

### AI assistance disclosure

Significant portions of this code were generated with AI assistance (Cursor/Claude). 
In particular, extracting the list of real metdata from dnf database on real rhel9 system and storing it as go fixture.The user:
- Provided requirements and architectural decisions
- Reviewed all generated code
- Corrected logic errors
- Requested specific refactoring (extracting helper functions)
- Validated test coverage and assertions

### CODE REVIEW

**Comment by vikin91** [comment]
This change is part of the following stack:

- #18190 ◀
    - #18191

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: pkg/fixtures/vmindexreport/generator.go
Comment (guzalv) [inline pkg/fixtures/vmindexreport/generator.go:1]: This generator is inside `pkg/fixtures`. On the one hand that is nice because it's next to the data it uses. On the other hand it breaks the pattern with the rest of fixtures in `pkg/fixtures` and it's not a fixture itself. Would it make sense to move this outside `fixtures`, and leave only e.g. `vm_package.go`?

File: pkg/fixtures/vmindexreport/generator.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+func NormalizeRPMVersion(version string) []int32 {
+	result := make([]int32, 10)
+
+	// Strip epoch prefix if present (e.g., "1:" in "1:1.54.0-3.el9_7")
+	if idx := strings.Index(version, ":"); idx != -1 {
+		version = version[idx+1:]
+	}
+
+	// Extract first 3 numeric sequences from the version string
+	matches := numericRegex.FindAllString(version, 3)
```
Comment (guzalv) [inline pkg/fixtures/vmindexreport/generator.go:38]: Some versions don't have a patch element, see this example from a real index report:

```json
{
        "id": 131,
        "name": "pigz",
        "version": "2.8-1.el9",
        ...
}
```

That would make this function return misleading data, but perhaps this is not a problem?

File: pkg/fixtures/vmindexreport/generator.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	packages     map[string]*v4.Package
+	repositories map[string]*v4.Repository
+	environments map[string]*v4.Environment_List
+}
+
+// selectPackageIndices returns a slice of indices into Rhel9Packages based on numRequested.
+// - numRequested <= 0: returns all indices sequentially [0, 1, ..., totalAvailable-1]
+// - numRequested < totalAvailable: randomly samples numRequested indices
+// - numRequested >= totalAvailable: uses all indices, then duplicates randomly to fill
+func selectPackageIndices(rng *rand.Rand, numRequested, totalAvailable int) []int {
```
Comment (guzalv) [inline pkg/fixtures/vmindexreport/generator.go:60]: I am considering whether changing the randomness based on how `numRequested` compares to `totalAvailable` is necessary. Now, if it is exactly `totalAvailable` we will always get the same values, if it's larger we will get some random values too, and if it's smaller we'll get completely random values.

How about we make this simpler and always return random data? e.g.:

```go
if numRequested <= 0 {
	numRequested = totalAvailable
}

indices := make([]int, numRequested)
for i := 0; i < numRequested; i++ {
	indices[i] = rng.Intn(totalAvailable)
}
return indices
```

If we want to ensure we get consistent data we can seed rng accordingly, and if we want to ensure non-repetition I'd propose doing it regardless of whether `numRequested` is lower or equal to `totalAvailable`, and forbidding requesting more packages than available. Perhaps 2 functions are better, if there's a need for such assurances.

File: pkg/fixtures/vmindexreport/generator.go [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+		for i := totalAvailable; i < numRequested; i++ {
+			indices[i] = rng.Intn(totalAvailable)
+		}
+		return indices
+	}
+}
+
+// buildRepositories creates real RHEL 9 repositories and synthetic ones if numRepos exceeds available.
+// Returns the repositories map and synthetic repo IDs (for assigning duplicated packages).
+// If less than `len(Rhel9Repositories)==2` are requested, then it always returns the two real repositories.
```
Comment (guzalv) [inline pkg/fixtures/vmindexreport/generator.go]:
```suggestion
// If less than `len(Rhel9Repositories)` are requested, then it always returns real repositories.
```

File: pkg/fixtures/vmindexreport/generator.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+func buildRepositories(numRepos int) (map[string]*v4.Repository, []string) {
+	totalRealRepos := len(Rhel9Repositories)
```
Comment (guzalv) [inline pkg/fixtures/vmindexreport/generator.go]: (this also applies to packages) I have a feeling that repos being rhel9-specific should be something encoded in the fixture, that then this function can use/ignore. It somehow feels weird to see rhel9 hardcoded here. How about something like this?

```python
repos = [{
    os: "rhel9",
    id: "rhel-9-for-x86_64-baseos-rpms",
    uri: "https://cdn.redhat.com/content/dist/rhel9/rhel-9-for-x86_64-baseos-rpms",
    cpe: "cpe:2.3:o:redhat:enterprise_linux:9:*:baseos:*:*:*:*:*"
}, {
    os: "rhel9",
    id: "rhel-9-for-x86_64-appstream-rpms",
    uri: "https://cdn.redhat.com/content/dist/rhel9/rhel-9-for-x86_64-appstream-rpms",
    cpe: "cpe:2.3:o:redhat:enterprise_linux:9:*:appstream:*:*:*:*:*"
}, {
    os: "rhel10",
    id: "rhel-10-for-x86_64-baseos-rpms",
    uri: "https://cdn.redhat.com/content/dist/rhel10/rhel-10-for-x86_64-baseos-rpms",
    cpe: "cpe:2.3:o:redhat:enterprise_linux:10:*:baseos:*:*:*:*:*"
}]
```

I won't insist if we want to accept that so far we only have rhel9 data and this is fine. But I'd rather accommodate other OSs from the get go.

File: pkg/fixtures/vmindexreport/generator.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+		var assignedRepoID string
+		if i < totalPkgs {
+			// Real package: keep original repo from fixture
+			assignedRepoID = pkg.Repo
+		} else if len(syntheticRepoIDs) > 0 {
+			// Duplicated package: assign to synthetic repos round-robin
+			assignedRepoID = syntheticRepoIDs[(i-totalPkgs)%len(syntheticRepoIDs)]
+		} else {
+			// No synthetic repos available, use original repo
```
Comment (guzalv) [inline pkg/fixtures/vmindexreport/generator.go:165]: Should we guard against this situation? i.e. require that there is always a synthetic repo if user requests more pkgs than available in real data?

Alternatively, if having duplicates here is fine, we could get rid of synthetic repos altogether.

**Review by janisz — APPROVED** [review]

File: pkg/fixtures/vmindexreport/generator.go
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:1]: You're right that it's not a "fixture" in the traditional sense. However, the generator is tightly coupled to the fixture data in this package. I'd prefer to keep it co-located for now and revisit if we add more generators or the package grows significantly. I am okay to move it if you feel strongly about this.

File: pkg/fixtures/vmindexreport/generator.go (hunk 1 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:38]: For load testing purposes, extracting the first 3 numeric components (2, 8, 1 in your example) is sufficient. The real scanner's version matching reads only the first two numbers to pre-filter as far as I can remember. The rest is matched using string version.
If all of those would be 0, then the string matching would be applied and result would be the same (but it would execute slightly longer).

File: pkg/fixtures/vmindexreport/generator.go (hunk 2 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:60]: To be honest, I think, it really does not matter. Important is only that scanner would treat the packages seriously and fire the SQL request against the database to check for vulns and compose the reply.
Scanner performance depends on whether a given package has many vulnerabilities or has none and that would be very difficult to control.

I can apply your suggestion if you feel strong about this.

File: pkg/fixtures/vmindexreport/generator.go (hunk 3 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go]: To be honest, I never planned adding more that those two, so I hardcoded it :)

File: pkg/fixtures/vmindexreport/generator.go (hunk 4 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go]: I see that your suggestions are based on a plan that we maybe would continue adding further things into the dataset for the packages. I didn't plan for it. I believe that we would use this code only for the current release and later only when we add new features that would need a scale check. When that happens, we can make it more generic, but for now we have a bigger fish to fry and I would prefer focus on it (preventing sensor and central from OOMing).

Of course if you have strong opinion about it, I can add it, but I would prefer to do this in a followup.

File: pkg/fixtures/vmindexreport/generator.go (hunk 5 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:165]: The synthetic repo fallback (line 165-167) exists to handle edge cases gracefully rather than panic. 

Regarding: avoiding synthetic data entirely — I agree this would be ideal for realistic load testing. However, expanding the fixture to 2000+ real packages would create extra work with minimal benefits to the feature. The current 509 packages cover the common case well, and synthetic repos only kick in for extreme load testing scenarios where we care about the size of a single grpc message. 
When I now think about it, ideally, we could drop the `numRepos` setting and always use only the two hardcoded ones.

**Comment by vikin91** [comment]
> * In `buildRepositories`, the behavior when `numRepos` is between 1 and `len(Rhel9Repositories)` may be surprising since it always returns all real repos (2) instead of honoring the requested count; consider either adjusting the logic or clarifying the function comment to match the actual behavior.

Updated the doc comment to clarify that all real repositories are always included when numRepos <= len(Rhel9Repositories) to ensure packages have valid repository references.

> * `MockDigestWithRegistry` in `vmindexreport/generator.go` is currently unused; consider removing it or wiring it into the code path where a digest with a registry prefix is needed to avoid dead constants.

This has been removed.


@sourcery-ai: This is load testing tooling, not production code. The existing tests verify the core functionality (report structure, package/repo counts, CPE parsing). Adding comprehensive edge case tests as suggested would be over-engineering for this use case. The generator's behavior is straightforward and deterministic with the seed parameter.

File: pkg/fixtures/vmindexreport/generator.go (hunk 4 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go]: I tried to get rid of rhel9 from var names in [ecc1fdb](https://github.com/stackrox/stackrox/pull/18190/commits/ecc1fdbfee9718a27f5edafa5e4b736243d47de6)

File: pkg/fixtures/vmindexreport/generator.go (hunk 3 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go]: I rephrased that comment, so the hardcoded 2 should be gone.

File: pkg/fixtures/vmindexreport/generator.go
Comment (guzalv) [reply pkg/fixtures/vmindexreport/generator.go:1]: I feel strongly about very few things, definitely this is not one of them: just a suggestion, but I'm confident you know better than me so I'm happy to leave it as is!

File: pkg/fixtures/vmindexreport/generator.go (hunk 1 above)
Comment (guzalv) [reply pkg/fixtures/vmindexreport/generator.go:38]: Thanks for the explanation. I would still add a comment that we know this is flawed but it's fine. But I leave that to your preference!

File: pkg/fixtures/vmindexreport/generator.go (hunk 2 above)
Comment (guzalv) [reply pkg/fixtures/vmindexreport/generator.go:60]: I would prefer that the function behaves in the same way (always random/always verbatim) regardless of `numRequested` v `totalAvailable`, but that's just for coherence. I understand it does not matter in this case and I agree we have more important things to do, so I am happy with this version if you prefer to keep it like this

File: pkg/fixtures/vmindexreport/generator.go (hunk 5 above)
Comment (guzalv) [reply pkg/fixtures/vmindexreport/generator.go:165]: Yes, I was suggesting to get rid of this particular edge case instead. This is anyway fine as-is, and I would definitely +1 using always the real repos

**Comment by vikin91** [comment]
The sensor integration tests failed with:

```
Installing kind...
/opt/hostedtoolcache/kind/v0.26.0/amd64/kind/bin ~/work/stackrox/stackrox
sha256sum: WARNING: 1 computed checksum did NOT match
kind-linux-amd64: FAILED
```
seeing this for the first time, so I am documenting it.

File: pkg/fixtures/vmindexreport/generator.go
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:1]: > I'm confident you know better than me

I am not so confident about this 😂  but thanks for the trust 👍

File: pkg/fixtures/vmindexreport/generator.go (hunk 1 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:38]: Thanks! I can add a comment if there is a rebase or another commit needed. If CI is happy with the current state, I would prefer merging as is.

File: pkg/fixtures/vmindexreport/generator.go (hunk 2 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:60]: I understand. Based on what you wrote about the preferences, I would propose to address that in one of the follow-ups. Would you be okay opening a PR for this?

File: pkg/fixtures/vmindexreport/generator.go (hunk 5 above)
Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:165]: Ok, let's drop the `numRepos` in the following PR. I am also open for adding more packages to the fixture. We just need to make sure that those are not a burden to maintain in the future.

Comment (vikin91) [reply pkg/fixtures/vmindexreport/generator.go:165]: @guzalv here is the followup PR for that: https://github.com/stackrox/stackrox/pull/18203 PTAL when you have time (this is not urgent)
